docs/source/tutorial/introduction.rst

```diff
@@ -0,0 +1,438 @@
+different from undirected graphs where edges have no notion of a direction to
+them. In retworkx the :class:`~retworkx.PyDiGraph` class is used to create
+directed graphs. For example:
+
+.. jupyter-execute::
+
+    import retworkx as rx
+    from retworkx.visualization import mpl_draw
+
+    path_graph = rx.generators.directed_path_graph(5)
+    mpl_draw(path_graph)
+
+In this example we created a 5 node directed path graph. This shows the
+directionality of the edges in the graph visualization with the arrow head
+pointing to the target node.
+
+Multigraphs
+===========
+
+By default all graphs in retworkx are multigraphs. This means that each
+graph object can contain parallel edges between nodes. However, you can set
+the ``multigraph`` argument to ``False`` on the :class:`~retworkx.PyGraph` and
+:class:`~retworkx.PyDiGraph` constructors when creating a new graph object will
+not allow you to add parallel edges. When ``multigraph`` is set to ``False`` if
+a method call is made that would add a parallel edge it will instead update the
+existing edge’s weight/data payload. For example:
+
+.. jupyter-execute::
+
+    graph = rx.PyGraph(multigraph=False)
+    graph.add_nodes_from(range(3))
+    graph.add_edges_from([(0, 1, 'A'), (0, 1, 'B'), (1, 2, 'C')])
+    mpl_draw(graph, with_labels=True, edge_labels=str)
+
+In this example when we attempted to add a parallel edge between nodes ``0``
+and ``1`` the edge data payload is updated from ``'A'`` to ``'B'``.
+
+Graph Generators and operations
+===============================
+
+In addition to constructing graphs one node and edge at a time, you can also
+create graphs using the :ref:`generator_funcs`, :ref:`random_generators`,
+and :ref:`graph-ops` to quickly generate graphs and/or apply different
+operations on the graph. For example:
+
+.. jupyter-execute::
+
+    lolipop_graph = rx.generators.lollipop_graph(4, 3)
+    mesh_graph = rx.generators.mesh_graph(4)
+    combined_graph = rx.cartesian_product(lolipop_graph, mesh_graph)[0]
+    mpl_draw(combined_graph)
+
+Additionally there are alternate constructors such as
+:meth:`~retworkx.read_edge_list` or :meth:`~retworkx.from_adjacency_matrix`
+for building graphs from files or other inputs. For example:
+
+.. jupyter-execute::
+
+    import tempfile
+
+    with tempfile.NamedTemporaryFile('wt') as fd:
+        path = fd.name
+        fd.write('0 1\n')
+        fd.write('0 2\n')
+        fd.write('0 3\n')
+        fd.write('1 2\n')
+        fd.write('2 3\n')
+        fd.flush()
+        graph = rx.PyGraph.read_edge_list(path)
+    mpl_draw(graph)
+
+Analyzing graphs
+================
+
+The structure of a graph :math:`G` can be analyzed using the graph algorithm
+functions. For example:
+
+.. jupyter-execute::
+
+    G = rx.PyGraph()
+    G.extend_from_edge_list([(0, 1), (0, 2)])
+    new_node = G.add_node("spam")
+    print(rx.connected_components(G))
+    degrees = []
+    for node in G.node_indices():
+        degrees.append(G.degree(node))
```

Comment:
What is the complexity of `G.degree()`? Constant time or linear in the degree itself (because of the need to traverse the linked list representing incident edges)?

Reply:
It is linear as to the number of edges on the node as it iterates over the edges list and return the count: https://github.com/Qiskit/retworkx/blob/main/src/graph.rs#L945-L955

In local testing I think the rust compiler can do a fair number of optimizations through the use of iterators and it's quite quick:

![degree_scaling](https://user-images.githubusercontent.com/2447371/147298061-cdaa68a4-9cb8-457a-bbc6-f891f682ad5e.png)

scaling up to 10^8 incident edge it only takes ~0.24 sec.

Generated with:
```python
import time

import retworkx as rx
import numpy as np
import matplotlib.pyplot as plt

sizes = np.logspace(0, 8, dtype=int)
print(sizes)
times = []
for size in sizes:
    g = rx.PyGraph()
    n = g.add_node(None)
    edge_list = [(n, i) for i in range(1, size + 1)]
    g.extend_from_edge_list(edge_list)
    start = time.time()
    deg = g.degree(n)
    stop = time.time()
    times.append(stop - start)
    print(deg)
print(times)
plt.plot(sizes, times)
plt.title("PyGraph.degree(node) runtime scaling")
plt.ylabel("Runtime (sec.)")
plt.xlabel("Number of incident edges (log scale)")
plt.savefig("/tmp/degree_scaling.png", dpi=900)
```

Reply:
I was just curious, and I was wondering if this is a weakness of the data structure you are using. I can't think of many applications where this would become a bottleneck though, and even if it would, it would be easy to just store the degrees within the graph data structure, for immediate access.